### PR TITLE
Check all arches in backend 

### PR DESCRIPTION
**Pull request by gretay-js**
The first commit in this PR adds Makefile build targets `make check_arch ARCH=<..>` and `make check_all_arches` for `backend` directory, with the same meaning as upstream, except that `check_all_arches` stops on first error (the behavior I've wanted whenever I used it before).  ARCH env variable is then used in dune rules in `backend` instead of the value of ARCH in the current configuration.
The last commit adds `check_all_arches` to CI as a separate workflow. It will be particularly useful with some form of #149 for catching Partial Match (warning 8).
The other commits fix various build errors.

### CODE REVIEW

File: Makefile.in [hunk 1]
```diff
@@ -92,8 +94,46 @@ hacking: ocaml-stage1-config.status stage0 \
...
+# This rule provides a quick way to check that machine-dependent
+# files compiles fine for a foreign architecture (passed as ARCH=xxx).
+# Based on stage1 target above.
+.PHONY: check_arch
+check_arch: ocaml-stage1-config.status stage0 \
+	config_files
+	cp ocaml-stage1-config.status ocaml/config.status
+	(cd ocaml && ./config.status)
+	@echo "========= CHECKING backend/$(ARCH) =============="
+	@rm -f $(ARCH_SPECIFIC)
```
Comment (mshinwell) [inline Makefile.in:116]: Please don't use `@` except for `echo`.  It makes debugging harder.

File: Makefile.in [hunk 2]
```diff
@@ -92,8 +94,46 @@ hacking: ocaml-stage1-config.status stage0 \
...
+	@rm -f backend/*.cm* backend/cfg/*.cm* backend/debug/*.cm*
```
Comment (mshinwell) [inline Makefile.in:117]: I don't understand this line -- this seems to be operating on the source directory?

File: Makefile.in [hunk 3]
```diff
@@ -92,8 +94,46 @@ hacking: ocaml-stage1-config.status stage0 \
...
+	PATH=$(stage0_prefix)/bin:$$PATH \
+	  $(dune) build --profile=release --build-dir=_build1 ocamloptcomp.cma
+	@rm -f $(ARCH_SPECIFIC)
+	@rm -f backend/*.cm* backend/cfg/*.cm* backend/debug/*.cm*
```
Comment (mshinwell) [inline Makefile.in:121]: Same comment as above.

File: .github/workflows/check_all_arches.yml [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+    - name: Configure Flambda backend (Closure mode)
+      working-directory: flambda_backend
+      run: |
+        ./configure \
+          --prefix=$GITHUB_WORKSPACE/_install \
+          --enable-middle-end=closure \
+          --with-dune=$GITHUB_WORKSPACE/dune/dune.exe
+
+    - name: Build all targets (Closure mode)
```
Comment (mshinwell) [inline .github/workflows/check_all_arches.yml:69]: This should say "Build check_all_arches" or similar.

File: backend/s390x/emit.mlp [hunk 5]
```diff
@@ -546,6 +548,7 @@ let emit_instr i =
         let instr = name_for_specific sop in
         `	{emit_string instr}	{emit_reg i.res.(0)}, {emit_reg i.arg.(0)}, {emit_reg i.arg.(1)}\n`
     | Lop (Iname_for_debugger _) -> ()
+    | Lop (Iprobe _ |Iprobe_is_enabled _) -> assert false
```
Comment (mshinwell) [inline backend/s390x/emit.mlp:551]: These should be `Misc.fatal_error "Not implemented"`, not `assert false`, I think?  The latter should be reserved for cases that are impossible to reach by inspection of nearby code, which isn't the case here.

**Review by mshinwell — CHANGES_REQUESTED** [review]

File: Makefile.in (hunk 1 above)
Comment (gretay-js) [reply Makefile.in]: fixed

File: Makefile.in (hunk 2 above)
Comment (gretay-js) [reply Makefile.in]: You are right. I've removed it. Dune knows to rebuild the backend directory when .ml files change.

File: Makefile.in (hunk 3 above)
Comment (gretay-js) [reply Makefile.in]: fixed.

File: backend/s390x/emit.mlp (hunk 5 above)
Comment (gretay-js) [reply backend/s390x/emit.mlp]: fixed.

**Review by mshinwell — APPROVED** [review]

**Pull request merged**